### ISSUE TITLE
Rethink reboot methodology

### ISSUE THREAD

**Issue by rbradford** [cloud-hypervisor/cloud-hypervisor#478]
Currently for reboot we tear down large parts of the VMM: in particular DeviceManager and CpuManager. This design decision was made to avoid issues with resetting the devices (in particular needing to implement VIRTIO device reset.) However as we look towards implementing hotplug of devices it would be better to be able to just destroy the vCPU threads and preserve the rest of the state.

**Comment by rbradford** [comment]
tl;dr I recommend we continue with our current approach and work out how to update our internal state from hotplugged devices

Currently when rebooting we destroy the `DeviceManager` and `CpuManager` and the whole VM (from KVM's perspective) too. Upon reboot we reuse the configuration and boot again as if it was a clean boot. The motivation behind this bug was to look into a less destructive solution that would allow us to keep the `DeviceManager` and thus all the devices alive and also making handling hotplug of devices easier.

I've done some experiments and here are my conclusions:

1. There is no way to destroy a vCPU from the KVM perspective, destroying the VM does destroy all the existing vCPUs on that VM but there is no vCPU destruction per-se.
2. Alternatively I prototyped keeping the vCPU threads alive and instead pausing them, calling `vcpu.configure()` again. This also requires reloading the kernel in the middle.
3. It occurs to me that we need to update our internal state based off the initial configuration so that the HTTP API gives the correct result, if we're managing our state for that then we might as use that for doing the reboot as we currently do.

**Comment by sboeuf** [comment]
@rbradford 
>It occurs to me that we need to update our internal state based off the initial configuration so that the HTTP API gives the correct result, if we're managing our state for that then we might as use that for doing the reboot as we currently do.

Well this sounds reasonable, as this will be the less intrusive method to support rebooting with the right amount of vCPU, memory, etc...

As discussed offline, one question remains for the vhost-user case, as a reboot will cause the socket to be destroyed, hence the communication with any vhost-user daemon will be closed.

**Comment by rbradford** [comment]
We've decided to stick with the current reboot methodology and will address the vhost-user issues separately.
